ffi/src/lib.rs

```diff
@@ -313,162 +568,365 @@ fn unwrap_conversion<T>(value: Option<T>, expected_type: ValueType) -> T {
+/// A destructor `typed_value_list_iter_destroy` is provided for releasing the memory for this
+/// pointer type.
 #[no_mangle]
 pub unsafe extern "C" fn values_iter(values: *mut Vec<TypedValue>) ->  *mut TypedValueIterator {
     let result = Box::from_raw(values);
     Box::into_raw(Box::new(result.into_iter()))
 }
 
+/// Returns the next value in the `iter` as a `ValueType`.
+/// If there is no value next value, a null pointer is returned.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `Vec<TypedValue>` is not dangling .
+///
+/// Callers are responsible for managing the memory for the return value.
+/// A destructor `typed_value_destroy` is provided for releasing the memory for this
+/// pointer type.
 #[no_mangle]
 pub unsafe extern "C" fn values_iter_next(iter: *mut TypedValueIterator) ->  *mut TypedValue {
     let iter = &mut *iter;
     iter.next().map_or(std::ptr::null_mut(), |v| Box::into_raw(Box::new(v)))
 }
 
+/// Returns the value at the provided `index` as a `ValueType`.
+/// If there is no value present at the `index`, a null pointer is returned.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `Vec<TypedValue>` is not dangling .
+///
+/// Callers are responsible for managing the memory for the return value.
+/// A destructor `typed_value_destroy` is provided for releasing the memory for this
+/// pointer type.
 #[no_mangle]
 pub unsafe extern "C" fn value_at_index(values: *mut Vec<TypedValue>, index: c_int) ->  *const TypedValue {
     let result = &*values;
     result.get(index as usize).expect("No value at index") as *const TypedValue
 }
 
-//as_long
+/// Returns the value of the `TypedValue` at `index` as a `long`.
+///
+/// # Panics
+///
+/// If the `ValueType` of the `TypedValue` is not `ValueType::Long`.
+/// If there is no value at `index`.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `Vec<TypedValue>` is not dangling .
 #[no_mangle]
 pub unsafe extern "C" fn value_at_index_as_long(values: *mut Vec<TypedValue>, index: c_int) ->  c_longlong {
     let result = &*values;
     let value = result.get(index as usize).expect("No value at index");
     unwrap_conversion(value.clone().into_long(), ValueType::Long)
 }
-// as ref
+
+/// Returns the value of the `TypedValue` at `index` as an `Entid`.
+///
+/// # Panics
+///
+/// If the `ValueType` of the `TypedValue` is not `ValueType::Ref`.
+/// If there is no value at `index`.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `Vec<TypedValue>` is not dangling .
 #[no_mangle]
 pub unsafe extern "C" fn value_at_index_as_entid(values: *mut Vec<TypedValue>, index: c_int) ->  Entid {
     let result = &*values;
     let value = result.get(index as usize).expect("No value at index");
     unwrap_conversion(value.clone().into_entid(), ValueType::Ref)
 }
 
-// as kw
+/// Returns the value of the `TypedValue` at `index` as a keyword `String`.
```

Comment:
`String` seems wrong here, since it's _not_ a Rust `String`.

Throughout, is there a reason we're not using Rust's `CString` (for owned C-strings) and `CStr` (for references to C-strings)?  I feel like I'm seeing a bunch of deviations from what I expect in the API, but I just don't know enough to be evaluate if those deviations are justified.

Enlighten me!